Added IterableExtensions.lastOrNull and deprecated last

LorenzoBettini opened the pull request:
Closes #2981

I have updated the occurrences in our code base, but NOT in the input files of our tests.

szarnekow commented on org.eclipse.xtext.xbase.lib.tests/src/org/eclipse/xtext/xbase/tests/lib/IterableExtensionsTest.java:
```diff
@@ -87,7 +87,7 @@ protected Integer findLast(Iterable<Integer> input, Function1<Integer, Boolean>
 
 	@Override
 	protected Integer last(Iterable<Integer> input) {
```
Can you please rename this method `last` also to `lastOrNull`?

LorenzoBettini replied:
@szarnekow this method is defined as abstract in the superclass `org.eclipse.xtext.xbase.tests.lib.BaseIterablesIteratorsTest.last(IterableOrIterator)` so I should rename it also in `IteratorExtensionsTest`.
I changed to `lastOrNull` only in `IterableExtensions`, NOT in `IteratorExtensions`.

szarnekow replied:
Yes, I saw that. I wonder if we should both keep in symmetry, though. Thoughts?

LorenzoBettini replied:
I also thought about that, but how do we "justify" the deprecation for `IteratorExtensions`? Of course, we could use the "symmetry" justification.

Also `head` could be changed accordingly.

Unfortunately, we are forced to change `IterableExtensions.last` only for the clash with Java 21 new API.

szarnekow replied:
So far I liked that the extension API for iterables and iterators was equivalent (as far as it goes).

LorenzoBettini replied:
Ok, then I'll add lastOrNull to IteratorExtensions and update this PR

LorenzoBettini commented:
@szarnekow I've added `lastOrNull` to `IteratorExtensions` as well, and updated also the tests according to https://github.com/eclipse/xtext/pull/2983#discussion_r1555733997

when the build is green, I'll merge
